tools/tcpstates.py

```diff
@@ -159,14 +159,19 @@
         ipv6_events.perf_submit(args, &data6, sizeof(data6));
     }
 
+    u64 ts = bpf_ktime_get_ns();
+    u64 *temp = last.lookup(&sk);
+    if (temp == NULL) {
+        last.insert(&sk, &ts);
+    }
     if (args->newstate == TCP_CLOSE) {
         last.delete(&sk);
```

Comment:
Maybe just do
```
  if (tsp)
    last.delete(&sk);
```
and there is no need to do lookup/insert. lookup has been done earlier and `update` is equivalent to `insert` for new items.

Reply:
I have tested it on my environment, (Linux kernel 5.8.0) , the verify process will fail if we use blow code 

```c
    u64 ts = bpf_ktime_get_ns();
    if (args->newstate == TCP_CLOSE) {
        if (tsp != NULL)
            last.delete(&sk);
    } else {
        last.update(&sk, &ts);
    }
```

Reply:
but this code below here is run correctly

```c
    u64 ts = bpf_ktime_get_ns();
    if (tsp == NULL) {
        last.insert(&sk, &ts);
    }
    if (args->newstate == TCP_CLOSE) {
        last.delete(&sk);
    } else {
        last.update(&sk, &ts);
    }
```

I think the verifier of eBPF has different restrict strategy in different Linux kernel version

Reply:
> I have tested it on my environment, (Linux kernel 5.8.0) , the verify process will fail if we use blow code
> 
> ```c
>     u64 ts = bpf_ktime_get_ns();
>     if (args->newstate == TCP_CLOSE) {
>         if (tsp != NULL)
>             last.delete(&sk);
>     } else {
>         last.update(&sk, &ts);
>     }
> ```

What kind of error did you see?